File: mergeSort.py
#! /usr/bin/python3.10
from numpy import random
import time
# Merge two sorted sublists `A[low … mid]` and `A[mid+1 … high]`
def merge(A, aux, low, mid, high):
 
    k = low
    i = low
    j = mid + 1
 
    # while there are elements in the left and right runs
    while i <= mid and j <= high:
 
        if A[i] <= A[j]:
            aux[k] = A[i]
            k = k + 1
            i = i + 1
        else:
            aux[k] = A[j]
            k = k + 1
            j = j + 1
 
    # copy remaining elements
    while i <= mid:
        aux[k] = A[i]
        k = k + 1
        i = i + 1
 
    # No need to copy the second half (since the remaining items
    # are already in their correct position in the auxiliary array)
 
    # copy back to the original list to reflect sorted order
    for i in range(low, high + 1):
        A[i] = aux[i]
 
 
# Sort list `A[low…high]` using auxiliary list aux
def mergesort(A, aux, low, high):
 
    # base case
    if high <= low:                     # if run size <= 1
        return
 
    # find midpoint
    mid = (low + ((high - low) >> 1))
 
    # recursively split runs into two halves until run size <= 1,
    # then merge them and return up the call chain
 
    mergesort(A, aux, low, mid)         # split/merge left half
    mergesort(A, aux, mid + 1, high)    # split/merge right half
 
    merge(A, aux, low, mid, high)       # merge the two half runs
 
 
# Function to check if `A` is sorted in ascending order or not
def isSorted(A):
 
    prev = A[0]
    for i in range(1, len(A)):
        if prev > A[i]:
            print("MergeSort Fails!!")
            return False
 
        prev = A[i]
 
    return True
 
 
# Implementation of merge sort algorithm in Python

A = random.randint(-5000, 5000, size=(10000))
aux = A.copy()
t = time.process_time()
# sort list `A` using auxiliary list `aux`
mergesort(A, aux, 0, len(A) - 1)
t = time.process_time() - t
if isSorted(A):
    print("CPU time: ",t)

File: MergeSort.java
import java.util.Arrays;

class MergeSort
{
    // Merge two sorted subarrays `arr[low … mid]` and `arr[mid+1 … high]`
    public static void merge(int[] arr, int[] aux, int low, int mid, int high)
    {
        int k = low, i = low, j = mid + 1;
 
        // while there are elements in the left and right runs
        while (i <= mid && j <= high)
        {
            if (arr[i] <= arr[j]) {
                aux[k++] = arr[i++];
            }
            else {
                aux[k++] = arr[j++];
            }
        }
 
        // copy remaining elements
        while (i <= mid) {
            aux[k++] = arr[i++];
        }
 
        // No need to copy the second half (since the remaining items
        // are already in their correct position in the auxiliary array)
 
        // copy back to the original array to reflect sorted order
        for (i = low; i <= high; i++) {
            arr[i] = aux[i];
        }
    }
 
    // Sort array `arr[low…high]` using auxiliary array `aux`
    public static void mergesort(int[] arr, int[] aux, int low, int high)
    {
        // base case
        if (high <= low) {        // if run size <= 1
            return;
        }
 
        // find midpoint
        int mid = (low + ((high - low) >> 1));
 
        // recursively split runs into two halves until run size <= 1,
        // then merge them and return up the call chain
 
        mergesort(arr, aux, low, mid);          // split/merge left half
        mergesort(arr, aux, mid + 1, high);     // split/merge right half
 
        merge(arr, aux, low, mid, high);        // merge the two half runs
    }
 
    // Function to check if arr is sorted in ascending order or not
    public static boolean isSorted(int[] arr)
    {
        int prev = arr[0];
        for (int i = 1; i < arr.length; i++)
        {
            if (prev > arr[i])
            {
                System.out.println("MergeSort Fails!!");
                return false;
            }
            prev = arr[i];
        }
 
        return true;
    }
 
    // Implementation of merge sort algorithm in Java
    public static void main(String[] args)
    {
        int[] arr = RandomUtils.randArray(1000, -5000, 5000);
        int[] aux = Arrays.copyOf(arr, arr.length);
        
 
        // sort array `arr` using auxiliary array `aux`
        var t = System.nanoTime();
        mergesort(arr, aux, 0, arr.length - 1);
        t = System.nanoTime() -t;
        double seconds = t * 0.00000001; 
        System.out.println("CPU time: " + seconds);
        // if (isSorted(arr)) {
        //     System.out.println(Arrays.toString(arr));
        // }
    }
}